### ISSUE TITLE
Why is calling model.build() required to load  weights for a subclassed tf.keras.Model

### ISSUE THREAD

**Issue by aaron276h** [tensorflow/tensorflow#44240]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): tf-nightly 
- TensorFlow version (use command below):2.4
- Python version: 3.7.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A 
- GPU model and memory:N/A

**Describe the current behavior**
This issue is related to models that subclass `tf.keras.Model`. When calling `model.load_weights("path.h5")` why is it required to have called `model.build()` beforehand? This [line](https://github.com/tensorflow/tensorflow/blob/8bfaaccad9594e215765cf927bc51c4bf8c0aad8/tensorflow/python/keras/engine/training.py#L2221) checks whether `model.built == True` and fails to load the model if it is not. `model.built` gets set to `True` if `model.build()` is called. However the doc-string for `model.build()` makes it seem as if it is an optional call:

```
This is to be used for subclassed models, which do not know at instantiation
time what their inputs look like.
This method only exists for users who want to call `model.build()` in a
standalone way (as a substitute for calling the model on real data to
build it). It will never be called by the framework (and thus it will
never throw unexpected errors in an unrelated workflow).
```

Is there a reason that `model.built` is checked when loading model weights?

I believe this issue appeared in TF2.3.

**Comment by Saduf2019** [comment]
@aaron276h 
Please share simple stand alone code to replicate issue faced or if possible share a colab gist with the issue.

**Comment by aaron276h** [comment]
@Saduf2019 Added a script the reproduces the error in TF 2.3+

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.python.keras.engine import data_adapter
from tensorflow.python.eager import backprop



def get_dataset():
    (mnist_images, mnist_labels), _ = \
        tf.keras.datasets.mnist.load_data(path='mnist.npz')

    dataset = tf.data.Dataset.from_tensor_slices(
        (tf.cast(mnist_images[..., tf.newaxis] / 255.0, tf.float32),
         tf.cast(mnist_labels, tf.int64))
    )
    dataset = dataset.repeat().shuffle(10000).batch(128)
    return dataset



class MNIST(keras.Model):
    def __init__(self):
        super(MNIST, self).__init__()
        self.mnist_model = keras.Sequential([
            tf.keras.Input((28, 28, 1,)),
            tf.keras.layers.Conv2D(32, [3, 3], activation='relu'),
            tf.keras.layers.Conv2D(64, [3, 3], activation='relu'),
            tf.keras.layers.MaxPooling2D(pool_size=(2, 2)),
            tf.keras.layers.Dropout(0.25),
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(128, activation='relu'),
            tf.keras.layers.Dropout(0.5),
            tf.keras.layers.Dense(10, activation='softmax')
        ])

    def call(self, inputs, training=None, mask=None):
        pass

    def compile(self, my_opt, loss):
        super(MNIST, self).compile()
        self.my_opt = my_opt
        self.loss_fn = loss

    def train_step(self, data):
        data = data_adapter.expand_1d(data)
        x, y, sample_weight = data_adapter.unpack_x_y_sample_weight(data)

        with backprop.GradientTape() as tape:
            y_pred = self.mnist_model(x, training=True)
            loss = self.loss_fn(y, y_pred, sample_weight)

        grads = tape.gradient(loss, self.trainable_variables)
        self.my_opt.apply_gradients(
            zip(grads, self.trainable_variables)
        )
        self.compiled_metrics.update_state(y, y_pred, sample_weight)
        metrics = {m.name: m.result() for m in self.metrics}
        metrics["loss"] = loss
        return metrics


def train():
    model = MNIST()
    optimizer = tf.keras.optimizers.Adam(0.001)

    model.compile(
        optimizer,
        loss=tf.losses.SparseCategoricalCrossentropy()
    )
    dataset = get_dataset()

    model.built = True
    model.fit(
        x=dataset.take(5),
        epochs=1
    )

    model.save_weights("model-weights.h5", save_format="h5")


def load():
    model = MNIST()
    optimizer = tf.keras.optimizers.Adam(0.001)
    model.compile(
        optimizer,
        loss=tf.losses.SparseCategoricalCrossentropy()
    )

    # If this line is commented out in TF 2.3+ the model fails to load.
    # model.built = True

    model.load_weights("model-weights.h5")
    print("Model loaded successfully.")


if __name__ == "__main__":
    train()
    load()
```

**Comment by Saduf2019** [comment]
@aaron276h 
Could you please refer to [this link](https://stackoverflow.com/questions/63658086/tensorflow-2-0-valueerror-while-loading-weights-from-h5-file) and let us know if it helps.

**Comment by aaron276h** [comment]
@Saduf2019 thanks for the link. It seems to indicate that saving subclassed model weights with `h5` format is no longer supported which is a bit strange as that use case worked perfectly fine if users just set `model.built = True`.

**Comment by ymodak** [comment]
The computation graph of custom objects such as subclassed models, custom metrics, custom layers, custom loss is not included in the saved file using `h5` format. 
See https://www.tensorflow.org/guide/keras/save_and_serialize#limitations
SavedModel formats saves the execution graph therefore we need not reinstantiate custom objects.

**Comment by aaron276h** [comment]
Thanks for the explanation @ymodak
